### PR TITLE
[9495] Make EMFILE exhauster run after usage of internal file descriptors 

### PR DESCRIPTION
**Pull request by yan12125**
## Scope and purpose

Glib-based main loops need a pair of sockets for handling signals [1],
while the file descriptor exhauster in tests runs before the main loop.

[1] https://gitlab.gnome.org/GNOME/pygobject/-/blob/3.42.0/gi/_ossighelper.py#L76

Closes https://twistedmatrix.com/trac/ticket/9495

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/9495
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [x] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #1729 from yan12125/9495-skip-non-working-glib-tests

Author: yan12125
Reviewer: <github_username>, <github_username_if_more_reviewers>
Fixes: ticket:9495

Skip test_closePeerOnEMFILE for Glib-based reactors

Glib-based main loops need a pair of sockets for handling signals [1],
while the file descriptor exhauster in tests runs before the main loop.

[1] https://gitlab.gnome.org/GNOME/pygobject/-/blob/3.42.0/gi/_ossighelper.py#L76
```

### CODE REVIEW

**Comment by yan12125** [comment]
Note that relevant tests are not run on CI until https://github.com/twisted/twisted/pull/1713 is merged.

**Comment by exarkun** [comment]
Could the test be adjusted so available file descriptors are exhausted after the reactor is started, instead of before?

**Comment by yan12125** [comment]
> Could the test be adjusted so available file descriptors are exhausted after the reactor is started, instead of before?

Good idea! I got a working yet very ugly test. If there is a better approach, I'm willing to give it another try.

**Comment by yan12125** [comment]
Failed tests from `3.7-noipv6-alldeps-withcov-posix` seem unrelated:

```
Error: 
Traceback (most recent call last):
Failure: twisted.internet.defer.TimeoutError: <twisted.test.test_process.PosixProcessPTYTests testMethod=test_signalTERM> (test_signalTERM) still running at 120.0 secs

twisted.test.test_process.PosixProcessPTYTests.test_signalTERM
===============================================================================
Error: 
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
Selectables:
<PTYProcess pid=2521 status=-1>

twisted.test.test_process.PosixProcessPTYTests.test_signalTERM
-------------------------------------------------------------------------------
```
(https://github.com/twisted/twisted/runs/6326471212)

File: src/twisted/internet/test/test_tcp.py
Comment (exarkun) [inline src/twisted/internet/test/test_tcp.py]: I'd probably spell this `IReactorTime.providedBy(reactor)`

File: src/twisted/internet/test/test_tcp.py
Comment (exarkun) [inline src/twisted/internet/test/test_tcp.py:1240]: I'd probably argue that `_glibbase.py` is violating the `IReactorCore` contract by calling `callWhenRunning` functions before it has gotten its signal handlers in place. :/

I don't really want to ask you to fix that for this PR (the problems looks gnarly to me, depending on how three different implementations of `_handleSignals` interact, at least) ... however I also think it is true that it sucks that any application code that really wants to be safe will have to duplicate this `callLater` logic instead of using `callWhenRunning`.  Fortunately, it's a really obscure cases where `callWhenRunning` breaks the interface, I guess.

I'd be okay with a compromise of describing this misbehavior of `_glibbase.py` in a new ticket so there's at least a slightly greater chance of someone fixing the underlying issue.  It would probably be good to link to that ticket from here and vice versa.

File: src/twisted/internet/test/test_tcp.py
Comment (exarkun) [inline src/twisted/internet/test/test_tcp.py]: Given the implementation details of `_glibbase.py` I think it might be safer to do `reactor.callLater(0, reactor.callLater, 0, inner)`.

Depending on how time passes and the exact scheduling of the thread running the reactor in this test, `callLater(0.01, ...` could end up being equivalent to `callLater(0, ...` which I suppose would still potentially interact poorly with `_globbase.py`'s re-installation of signal handlers.  The double `callLater(0, ...` should avoid this because no matter how time passes, the first `callLater(0, ...` here and the `callLater(0, ...` in `_glibbase.py` will happen in the same reactor iteration and only after that can the second `callLater(0, ...` here happen.

**Review by exarkun — CHANGES_REQUESTED** [review]
Thanks!  I left some comments inline.  I think there's some unfortunately subtle stuff going on here, but maybe not that needs to block merging this PR on a lot more work.  Please let me know if I can clarify any of my inline comments.  Thanks again!

**Comment by exarkun** [comment]
> Failed tests from `3.7-noipv6-alldeps-withcov-posix` seem unrelated:
> 
> ```
> Error: 
> Traceback (most recent call last):
> Failure: twisted.internet.defer.TimeoutError: <twisted.test.test_process.PosixProcessPTYTests testMethod=test_signalTERM> (test_signalTERM) still running at 120.0 secs
> 
> twisted.test.test_process.PosixProcessPTYTests.test_signalTERM
> ===============================================================================
> Error: 
> Traceback (most recent call last):
> Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
> Selectables:
> <PTYProcess pid=2521 status=-1>
> 
> twisted.test.test_process.PosixProcessPTYTests.test_signalTERM
> -------------------------------------------------------------------------------
> ```
> 
> (https://github.com/twisted/twisted/runs/6326471212)

Yes.  Those look like familiar spurious intermittent failures. :(

File: src/twisted/internet/test/test_tcp.py
Comment (yan12125) [reply src/twisted/internet/test/test_tcp.py]: Thanks, that's cleaner!

File: src/twisted/internet/test/test_tcp.py
Comment (yan12125) [reply src/twisted/internet/test/test_tcp.py]: That makes sense. The `reactor.callLater(0, reactor.callLater, 0, inner)` trick is brilliant :)

File: src/twisted/internet/test/test_tcp.py
Comment (yan12125) [reply src/twisted/internet/test/test_tcp.py:1240]: Yeah that sounds unusual. I don't think I can fix this in the short term, so I created a ticket as a reminder: https://twistedmatrix.com/trac/ticket/10342

**Comment by yan12125** [comment]
Thanks for reviewing! Explanations are clear and suggestions are useful :)

**Review by adiroiban — APPROVED** [review]
Look good. Thanks

**Comment by adiroiban** [comment]
So it looks like this PR is blocked as @exarkun needs to re-review this PR.

In theory, I can dismiss that review and merge, but I prefer to get a confirmation.

I hope this will be re-review soon and we can land it.

Thank you guys.

**Comment by yan12125** [comment]
Oh git commits become messy. I squashed them and drop commit messages about skipping tests. There are no code changes: https://github.com/twisted/twisted/compare/c42e06edd609d3325055cc09107830166d7efcc4..896a1d1bb039a06465d56c4dbefb5cc689d890ac

Meanwhile, I updated pull request title and description to reflect the current status.

**Review by exarkun — APPROVED** [review]
Thanks!

**Pull request merged**
